### ISSUE TITLE
`FROM x@\<newline>sha256:...`: `expecting no '@' or a valid digest hash`

### ISSUE THREAD

**Issue by kbulygin** [hadolint/hadolint#203]
- [x] This is a bug report
- [ ] This is a feature request
- [x] I searched existing issues before opening this one

### Steps to reproduce the behavior

**Output of `hadolint --version`:**

```bash
Haskell Dockerfile Linter v1.6.6-0-g9658652
```

**Dockerfile**

Case 1:
```Dockerfile
FROM python@\
sha256:b133ed1865bd7470239630c13e5834470f1c31673881924924643cefa461e9fb
...
```

Case 2:
```Dockerfile
FROM python\
@sha256:b133ed1865bd7470239630c13e5834470f1c31673881924924643cefa461e9fb
...
```

### Expected behavior

No warnings. The backslash in inserted to keep lines at no more than 80 characters. `docker build` (`Docker version 18.03.0-ce, build 0520e24`) parses `<before><backslash><newline><after>` correctly as `<before><after>`.

### Actual behavior

Case 1: `Dockerfile:1:13 unexpected '@' expecting no '@' or a valid digest hash (example: python@a3f42f2de)`
Case 2: `Dockerfile:1:13 unexpected "@" expecting space or "AS"`

It seems that `hadolint` parses `<before><backslash><newline><after>` as `<before><token separator><after>`. (Backslash handling is not part of text preprocessing, so `<before><after>` is not treated as a single token.)

**Comment by lorenzo** [comment]
Is that even allowed? The new line breaks in dockerfile are quite arbitrary. There are other places where breaking the line is not allowed at all. 

What hadolint does is that any backslash followed by a new line is treated as a single white space. 

This could take a long time to fix, if it actually gets fixed due to to the complexity it introduces. So I suggest using the workaround of not breaking the line for the time being.

**Comment by lorenzo** [comment]
@kbulygin actually, this particular case seems easier to fix than I though. I expect having a fix for it during this week.

Thanks for reporting!

**Comment by kbulygin** [comment]
@lorenzo Yes, it is allowed: https://docs.docker.com/engine/reference/builder/#escape. The term "escaped newline" is illustrated at https://docs.docker.com/engine/reference/builder/#run:

> In the shell form you can use a \ (backslash) to continue a single RUN instruction onto the next line. For example, consider these two lines:
> ```Dockerfile
> RUN /bin/bash -c 'source $HOME/.bashrc; \
> echo $HOME'
> ```
> Together they are equivalent to this single line:
> ```Dockerfile
> RUN /bin/bash -c 'source $HOME/.bashrc; echo $HOME'
> ```

Note that the number of spaces is the same after preprocessing.

Dockerfile syntax is consistent here with C (which probably introduced the syntax; newline elimination happens during preprocessing, try `cpp(1)` on files with backslashes), bash (try to type `l<backslash><newline>s`), recipes in Makefiles, and so on.

**Comment by kbulygin** [comment]
@lorenzo Thanks for reacting and for the wonderful linter!
